File: Trees/pre-in-post_order-traversal.py
# Traverse binary Tree
# if print root.data, then it will print next after a space due to ','

class Node:
    def __init__(self,data):
        self.data = data
        self.left = None
        self.right = None

class Tree:
    def __init__(self):
        self.currTree = None
    
    def addNode(self,data):
        if self.currTree == None:
            self.currTree =  Node(data)
        else:
            self._addNode(data,self.currTree)
        return self.currTree
    
    def _addNode(self,data,node):
        if data < node.data:
            if node.left != None:
                self._addNode(data,node.left)
            else:
                node.left = Node(data)
        else:
            if node.right != None:
                self._addNode(data,node.right)
            else:
                node.right = Node(data)   

    def PreOrderPrintNode(self,node):
        if node != None:
            print node.data
            self.PreOrderPrintNode(node.left)           
            self.PreOrderPrintNode(node.right)

    def InOrderPrintNode(self,node):
        if node != None:
            self.InOrderPrintNode(node.left)
            print node.data
            self.InOrderPrintNode(node.right)
    
    def PostOrderPrintNode(self,node):
        if node != None:
            self.PostOrderPrintNode(node.left)            
            self.PostOrderPrintNode(node.right)
            print node.data

if __name__ == "__main__":
    tr = Tree()
    tr.addNode(20)
    tr.addNode(3)
    node = tr.addNode(1)
    tr.InOrderPrintNode(node)

File: Trees/pre-in-post_order-traversal.js
//Traverse binary Tree

class Node {
    constructor(data) {
        this.data = data;
        this.left = null;
        this.right = null;
    }
}

class Tree {
    constructor() {
        this.currTree = null;
    }

    addNode(data) {
        if (this.currTree == null) {
            this.currTree = new Node(data);
        } else {
            this._addNode(data, this.currTree);
        }
        return this.currTree;
    }

    _addNode(data, node) {
        if (data < node.data) {
            if (node.left != null) {
                this._addNode(data, node.left);
            } else {
                node.left = new Node(data);
            }

        } else {
            if (node.right != null) {
                this._addNode(data, node.right);
            } else {
                node.right = new Node(data);
            }
        }
    }

    printInOrder(node) {
        if (node != null) {
            this.printInOrder(node.left);
            console.log(node.data);
            this.printInOrder(node.right);
        }
    }

    printPreOrder(node) {
        if (node != null) {
            console.log(node.data);
            this.printPreOrder(node.left);
            this.printPreOrder(node.right);
        }
    }

    printPostOrder(node) {
        if (node != null) {
            this.printPostOrder(node.left);
            this.printPostOrder(node.right);
            console.log(node.data);
        }
    }
}


var tr = new Tree();
tr.addNode(20);
tr.addNode(3);
var node = tr.addNode(1);
tr.printInOrder(node);